File: yurt-convention-api/src/main/java/com/github/liulus/yurt/convention/data/DefaultPage.java
package com.github.liulus.yurt.convention.data;

import java.util.Collections;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/9
 */
public class DefaultPage<T> implements Page<T> {

    private int pageNum;
    private int pageSize;
    private List<? extends T> results;
    private long totalRecords;

    public DefaultPage(int pageNum, int pageSize, List<? extends T> results, long totalRecords) {
        this.pageNum = pageNum;
        this.pageSize = pageSize;
        this.results = results;
        this.totalRecords = totalRecords;
    }

    public DefaultPage() {
        this(1, 20, Collections.emptyList(), 0);
    }

    public DefaultPage(List<? extends T> results) {
        this(1, 20, results, results.size());
    }

    public DefaultPage(Pageable pageable, List<? extends T> results, int totalRecords) {
        this(pageable.getPageNum(), pageable.getPageSize(), results, totalRecords);
    }


    @Override
    public int getPageNum() {
        return pageNum;
    }

    public void setPageNum(int pageNum) {
        this.pageNum = pageNum;
    }

    @Override
    public int getPageSize() {
        return pageSize;
    }

    public void setPageSize(int pageSize) {
        this.pageSize = pageSize;
    }

    @Override
    public List<? extends T> getResults() {
        return results == null ? Collections.emptyList() : results;
    }

    public void setResults(List<? extends T> results) {
        this.results = results;
    }

    @Override
    public long getTotalRecords() {
        return totalRecords;
    }

    public void setTotalRecords(long totalRecords) {
        this.totalRecords = totalRecords;
    }

    @Override
    public int getTotalPages() {
        return (int) ((totalRecords - 1) / pageSize + 1);
    }

    @Override
    public <R> Page<T> map(Function<T, R> transformer) {
        return new DefaultPage(pageNum, pageSize,
                results.stream().map(transformer).collect(Collectors.toList()), totalRecords);
    }
}

File: yurt-convention-api/src/main/kotlin/com/github/liulus/yurt/convention/data/DefaultPage.kt
package com.github.liulus.yurt.convention.data


/**
 *
 * @author liulu
 * @version V1.0
 * @since 2020/10/29
 */
data class DefaultPage<T>(
        override var pageNum: Int,
        override var pageSize: Int,
        override var results: List<T>,
        override var totalRecords: Int,
) : Page<T> {

    constructor() : this(PageQuery.DEFAULT_PAGE_NUM, PageQuery.DEFAULT_PAGE_SIZE, emptyList(), 0)

    constructor(results: List<T>) : this(PageQuery.DEFAULT_PAGE_NUM, PageQuery.DEFAULT_PAGE_SIZE, results, results.size)

    constructor(pageable: Pageable, results: List<T>, totalRecords: Int)
            : this(pageable.pageNum, pageable.pageSize, results, totalRecords)


    override val totalPages: Int
        get() = (totalRecords - 1) / pageSize + 1


    override fun <S> map(transformer: (T) -> S): Page<S> {
        return DefaultPage(pageNum, pageSize, results.asSequence().map(transformer).toList(), totalRecords)
    }


}